File: src/main/java/hara/lang/data/Queue.java
package hara.lang.data;

import java.util.ArrayList;
import java.util.Iterator;

import hara.lang.base.*;

public interface Queue<E> extends
	I.Coll<E>,
	Data.LinearType<E>,
	Data.SequentialLookupType<E> {

	public static int MAX_LENGTH = 1024;
	
	public interface Base<E> extends Queue<E> {
		public int _size();
		public int _offset();

		public Vector<E> _head();
		public Vector<E> _tail();
		public List<Vector<E>> _buffer();

		@Override
		default E peekFirst() {
			if(_size() == 0) { return null; }
			return _head().peekFirst();
		}
		
		@Override
		default E peekLast() {	
			if(_size() == 0) { return null; }
			if(_tail().count() > 0) {
				return _tail().peekLast();
			} else if (_buffer().count() == 0) {
				return _head().peekLast();
			} else {
				return _buffer().peekLast().peekLast();
			}
		}

		@SuppressWarnings({ "rawtypes", "unchecked" })
		@Override
		default Iterator<E> iterator() {
			if(_size() == 0) { return (Iterator<E>) It.emptyIterator(); };
			ArrayList<Iterator> all = new ArrayList<Iterator>();
			all.add(_head().iterator());
			It.reduce(_buffer().iterator(), all, (arr, v) -> {
				arr.add(v.iterator());
				return all;
			});
			all.add(_tail().iterator());
			return It.concat(all.iterator());
		}

		@Override
		default E nth(long i) {
			if (i >= 0 && i < _size()) {
				var space = MAX_LENGTH - _offset();
				if(i < space) {
					return _head().nth(_offset() + i);
				} else {
					int row = (int) ((i - space) % MAX_LENGTH);
					int col = (int) ((i - space) / MAX_LENGTH);
					
					if(row < _buffer().count()) {
						return _buffer().nth(row).nth(col);
					} else {
						return _tail().nth(col);
					}
				}
			} else {
				throw new Ex.OutOfBounds();
			}
		}

		@Override
		default long count() {
			return _size();
		}
		
		@Override
		default String startString() {
			return "[";
		}

		@Override
		default String endString() {
			return "]";
		}
	}
	
	public class Mutable<E> extends Data.RefType.MT 
		implements Base<E>, I.ToPersistent {
		int _size;
		int _offset;

		Vector<E> _head;
		Vector<E> _tail;
		List<Vector<E>> _buffer;

		@SuppressWarnings("unchecked")
		public Mutable(I.Metadata meta) {
			super(meta);
			_size = 0;
			_offset = 0;
			_head = Vector.Mutable.empty(null);
			_tail = Vector.Mutable.empty(null);
			_buffer = new List.Mutable<Vector<E>>(null, List.DEFAULT_CAPACITY);
		}


		public Mutable(I.Metadata meta, int size, int offset, Vector<E> head, Vector<E> tail,
				List<Vector<E>> buffer) {
			super(meta);
			_size = size;
			_offset = offset;
			
			_head = head;
			_tail = tail;
			_buffer = buffer;
		}

		@SuppressWarnings("unchecked")
		public static <E> Mutable<E> from(I.Metadata meta, E... objs) {
			var vec = Mutable.empty(meta);
			return Arr.reduce((v, e) -> v.pushLast(e), vec, objs);
		}
		
		@SuppressWarnings({"rawtypes"})
		public static Mutable empty (I.Metadata meta) {
			return new Mutable(meta);
		}


		public static <E> Mutable<E> into(Iterator<E> it) {
			return into(new Mutable<E>(null), it);
		}
		
		public static <E> Mutable<E> into(Mutable<E> coll, Iterator<E> it) {
			return It.reduce(it, coll, (m, e) -> m.pushLast(e));
		}

		@SuppressWarnings({ "unchecked"})
		@Override
		public Mutable<E> pushLast(E e) {
			var space = MAX_LENGTH - _offset;
			if(_size < space) {
				_head.conj(e);
			} else {
				_tail.conj(e);
				if(_tail.count() < MAX_LENGTH) {
					
				} else {
					_buffer.conj(_tail);
					_tail = Vector.Mutable.empty(null);
				}
			}
			_size++;
			return this;
		}
		
		@Override
		public Mutable<E> popFirst() {
			if( _size == 0) { return this; }
			
			_offset++;
			_size--;
			if(_offset < MAX_LENGTH) {
				return this;
			} else {
				_offset = 0;
				if(_buffer.count() == 0) {
					_head = _tail;
				} else {
					_head = _buffer.peekFirst();
					_buffer.popFirst();
				}
			}
			return this;
		}

		
		@SuppressWarnings({ "unchecked", "rawtypes" })
		@Override
		public Mutable<E> popLast() {
			
			if(_size == 0) { return this; }
			
			_size--;
			if(_tail.count() > 0) {
				_tail.popLast();
			} else if (_buffer.count() == 0) {
				_head.popLast();
			} else {
				_tail = _buffer.peekLast();
				_buffer = (List)_buffer.popLast();
			}
			return this;
		}
		

		@SuppressWarnings({ "unchecked" })
		@Override
		public Mutable<E> empty() {
			_size = 0;
			_offset = 0;
			_head = (Vector<E>) _head.empty();
			_tail = (Vector<E>) _tail.empty();
			_buffer = (List<Vector<E>>) _buffer.empty();
			return this;
		}

		@Override
		public int _size() {
			return _size;
		}

		@Override
		public int _offset() {
			return _offset;
		}

		@Override
		public Vector<E> _head() {
			return _head;
		}

		@Override
		public Vector<E> _tail() {
			return _tail;
		}

		@Override
		public List<Vector<E>> _buffer() {
			return _buffer;
		}
		@SuppressWarnings({ "unchecked", "rawtypes" })
		@Override
		public Standard<E> toPersistent() {
			var head = ((Vector.Mutable<E>)_head).toPersistent();
			var tail = ((Vector.Mutable<E>)_tail).toPersistent();
			var buff = ((List.Mutable)_buffer).toPersistent();
			return new Standard<E>(_meta, _size, _offset, head, tail, buff);
		}		
	}

	public class Standard<E> extends Data.RefType.PT 
		implements Base<E>, I.ToMutable {
		final int _size;
		final int _offset;

		final Vector<E> _head;
		final Vector<E> _tail;
		final List<Vector<E>> _buffer;
		
		// STATIC
		public final static Standard<Object> EMPTY = new Standard<>(null);

		@SuppressWarnings("unchecked")
		public Standard(I.Metadata meta) {
			super(meta);
			_size = 0;
			_offset = 0;
			_head = (Vector<E>) Vector.Standard.EMPTY;
			_tail = (Vector<E>) Vector.Standard.EMPTY;
			_buffer = List.Standard.EMPTY;
					
		}

		private Standard(I.Metadata meta, int size, int offset, Vector<E> head, Vector<E> tail,
				List<Vector<E>> buffer) {
			super(meta);
			_size = size;
			_offset = offset;
			
			_head = head;
			_tail = tail;
			_buffer = buffer;
		}
		
		@SuppressWarnings("unchecked")
		public static <E> Standard<E> empty(I.Metadata meta) {
			Standard<E> ret = (Standard<E>) EMPTY;
			return (meta == null) ? ret : ret.withMeta(meta);
		}

		@SuppressWarnings("unchecked")
		public static <E> Standard<E> from(I.Metadata meta, E... objs) {
			return Mutable.from(meta, objs).toPersistent();
		}
		
		public static <E> Standard<E> into(Iterator<E> it) {
			return Mutable.into(it).toPersistent();
		}
		
		@SuppressWarnings("unchecked")
		public static <E> Standard<E> into(Standard<E> coll, Iterator<E> it) {
			return Mutable.into(coll.toMutable(), it).toPersistent();
		}

		@SuppressWarnings({ "unchecked", "rawtypes" })
		@Override
		public Standard<E> pushLast(E e) {
			var space = MAX_LENGTH - _offset;
			int nsize = _size + 1;
			if(_size < space) {
				var nhead = (Vector) _head.conj(e);
				return new Standard<E>(_meta, nsize, _offset, nhead, _tail, _buffer);
			} else {
				var ntail = (Vector<E>) _tail.conj(e);
				if(ntail.count() < MAX_LENGTH) {
					return new Standard<E>(_meta, nsize, _offset, _head, ntail, _buffer);
				} else {
					return new Standard<E>(_meta, nsize, _offset, _head, 
							(Vector<E>) Vector.Standard.EMPTY,
							(List) _buffer.conj(ntail));
				}
			}	
		}
		
		@SuppressWarnings({ "unchecked", "rawtypes" })
		@Override
		public Standard<E> popFirst() {
			if( _size == 0) { return this; }
			var noffset = _offset + 1;
			if(noffset < MAX_LENGTH) {
				return new Standard<E>(_meta, _size - 1, noffset, _head, _tail, _buffer);
			} else {
				if(_buffer.count() == 0) {
					return new Standard<E>(_meta, _size - 1, 0, _tail,
							(Vector<E>) Vector.Standard.EMPTY, _buffer);
				} else {
					var nhead = _buffer.peekFirst();
					return new Standard<E>(_meta, _size - 1, 0, nhead,
							_tail, (List)_buffer.popFirst());
				}
			}
		}

		
		@SuppressWarnings({ "unchecked", "rawtypes" })
		@Override
		public Standard<E> popLast() {
			
			if(_size == 0) { return this; }
			var nsize = _size - 1;
			if(_tail.count() > 0) {
				return new Standard<E>(_meta, nsize, _offset, 
						_head,(Vector) _tail.popLast(), _buffer);
			} else if (_buffer.count() == 0) {
				return new Standard<E>(_meta, nsize, _offset, 
						(Vector) _head.popLast(), _tail, _buffer);
			} else {
				var ntail = _buffer.peekLast();
				return new Standard<E>(_meta, _size - 1, _offset, _head,
						ntail, (List)_buffer.popLast());
			}
		}
		
		@Override
		public Standard<E> withMeta(I.Metadata meta) {
			return (meta == _meta) ? this : new Standard<E>(meta, _size, _offset, _head, _tail, _buffer);
		}

		@SuppressWarnings({ "rawtypes", "unchecked" })
		@Override
		public Standard<E> empty() {
			return new Standard(_meta, 0, 0, 
					Vector.Standard.EMPTY,
					Vector.Standard.EMPTY,
					List.Standard.EMPTY);
		}

		@Override
		public int _size() {
			return _size;
		}

		@Override
		public int _offset() {
			return _offset;
		}

		@Override
		public Vector<E> _head() {
			return _head;
		}

		@Override
		public Vector<E> _tail() {
			return _tail;
		}

		@Override
		public List<Vector<E>> _buffer() {
			return _buffer;
		}

		@Override
		@SuppressWarnings({ "rawtypes", "unchecked" })
		public Mutable toMutable() {
			var head = ((Vector.Standard<E>)_head).toMutable();
			var tail = ((Vector.Standard<E>)_tail).toMutable();
			var buff = ((List.Standard)_buffer).toMutable();
			return new Mutable<E>(_meta, _size, _offset, head, tail, buff);
		}

	}
}

File: wasm/src/lang/data/queue.rs
use crate::lang::data::{List, Vector};
use crate::lang::protocol::{
    IConj, ICount, IEmpty, IMetadata, IMutable, INth, IPeekFirst, IPeekLast, IPersistent,
    IPopFirst, IPopLast, IPushLast, IToMutable, IToPersistent,
};
use std::rc::Rc;

pub const MAX_LENGTH: usize = 1024;

#[derive(Debug, Clone)]
pub struct Standard<E> {
    metadata: Option<Rc<crate::lang::data::Metadata>>,
    size: usize,
    offset: usize,
    head: Vector<E>,
    tail: Vector<E>,
    buffer: List<Vector<E>>,
}

impl<E: Clone> Default for Standard<E> {
    fn default() -> Self {
        Self {
            metadata: None,
            size: 0,
            offset: 0,
            head: Vector::new(),
            tail: Vector::new(),
            buffer: List::new(),
        }
    }
}
impl<E: Clone> Standard<E> {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn len(&self) -> usize {
        self.size
    }
    pub fn is_empty(&self) -> bool {
        self.size == 0
    }
    pub fn peek_first(&self) -> Option<&E> {
        if self.size == 0 {
            None
        } else {
            self.head.get(self.offset)
        }
    }
    pub fn peek_last(&self) -> Option<&E> {
        if self.size == 0 {
            None
        } else if !self.tail.is_empty() {
            self.tail.get(self.tail.len() - 1)
        } else if !self.buffer.is_empty() {
            self.buffer
                .get(self.buffer.len() - 1)
                .and_then(|v| v.get(v.len() - 1))
        } else {
            self.head.get(self.head.len() - 1)
        }
    }
    pub fn get(&self, index: usize) -> Option<&E> {
        if index >= self.size {
            return None;
        }
        let head_remaining = self.head.len().saturating_sub(self.offset);
        if index < head_remaining {
            return self.head.get(self.offset + index);
        }
        let after = index - head_remaining;
        let chunk = after / MAX_LENGTH;
        let column = after % MAX_LENGTH;
        if chunk < self.buffer.len() {
            self.buffer.get(chunk).and_then(|v| v.get(column))
        } else {
            self.tail.get(column)
        }
    }
    pub fn push_last(&self, value: E) -> Self {
        let space = MAX_LENGTH - self.offset;
        if self.size < space {
            return Self {
                size: self.size + 1,
                head: self.head.push_last(value),
                ..self.clone()
            };
        }
        let tail = self.tail.push_last(value);
        if tail.len() < MAX_LENGTH {
            Self {
                size: self.size + 1,
                tail,
                ..self.clone()
            }
        } else {
            Self {
                size: self.size + 1,
                tail: Vector::new(),
                buffer: self.buffer.push_last(tail),
                ..self.clone()
            }
        }
    }
    pub fn pop_first_value(&self) -> Self {
        if self.size == 0 {
            return self.clone();
        }
        let offset = self.offset + 1;
        if offset < MAX_LENGTH {
            return Self {
                size: self.size - 1,
                offset,
                ..self.clone()
            };
        }
        if self.buffer.is_empty() {
            Self {
                metadata: self.metadata.clone(),
                size: self.size - 1,
                offset: 0,
                head: self.tail.clone(),
                tail: Vector::new(),
                buffer: self.buffer.clone(),
            }
        } else {
            Self {
                size: self.size - 1,
                offset: 0,
                head: self.buffer[0].clone(),
                buffer: self.buffer.pop_first_value(),
                ..self.clone()
            }
        }
    }
    pub fn pop_last_value(&self) -> Self {
        if self.size == 0 {
            return self.clone();
        }
        if !self.tail.is_empty() {
            Self {
                size: self.size - 1,
                tail: self.tail.pop_last_value().expect("nonempty tail"),
                ..self.clone()
            }
        } else if self.buffer.is_empty() {
            Self {
                size: self.size - 1,
                head: self.head.pop_last_value().unwrap_or_else(Vector::new),
                ..self.clone()
            }
        } else {
            Self {
                size: self.size - 1,
                tail: self.buffer[self.buffer.len() - 1].clone(),
                buffer: self.buffer.pop_last_value(),
                ..self.clone()
            }
        }
    }
    pub fn iter(&self) -> impl Iterator<Item = &E> {
        (0..self.size).map(|i| self.get(i).expect("valid queue index"))
    }
}
impl<E: Clone> FromIterator<E> for Standard<E> {
    fn from_iter<T: IntoIterator<Item = E>>(iter: T) -> Self {
        iter.into_iter().fold(Self::new(), |q, v| q.push_last(v))
    }
}
impl<E: Clone + PartialEq> PartialEq for Standard<E> {
    fn eq(&self, other: &Self) -> bool {
        self.size == other.size && self.iter().eq(other.iter())
    }
}
impl<E: Clone> ICount for Standard<E> {
    fn count(&self) -> usize {
        self.size
    }
}
impl<E: Clone> INth<E> for Standard<E> {
    fn nth(&self, index: usize) -> Option<&E> {
        self.get(index)
    }
}
impl<E: Clone> IPeekFirst<E> for Standard<E> {
    fn peek_first(&self) -> Option<E> {
        Standard::peek_first(self).cloned()
    }
}
impl<E: Clone> IPeekLast<E> for Standard<E> {
    fn peek_last(&self) -> Option<E> {
        Standard::peek_last(self).cloned()
    }
}
impl<E: Clone> IPushLast<E> for Standard<E> {
    type Output = Self;
    fn push_last(&self, value: E) -> Self {
        Standard::push_last(self, value)
    }
}
impl<E: Clone> IConj<E> for Standard<E> {
    type Output = Self;
    fn conj(&self, value: E) -> Self {
        self.push_last(value)
    }
}
impl<E: Clone> IPopFirst for Standard<E> {
    type Output = Self;
    fn pop_first(&self) -> Self {
        self.pop_first_value()
    }
}
impl<E: Clone> IPopLast for Standard<E> {
    type Output = Self;
    fn pop_last(&self) -> Self {
        self.pop_last_value()
    }
}
impl<E: Clone> IEmpty for Standard<E> {
    type Output = Self;
    fn empty(&self) -> Self {
        Self::new().with_meta(self.metadata.clone())
    }
}
impl<E: Clone> IMetadata for Standard<E> {
    type Metadata = Rc<crate::lang::data::Metadata>;
    fn meta(&self) -> Option<&Self::Metadata> {
        self.metadata.as_ref()
    }
    fn with_meta(&self, metadata: Option<Self::Metadata>) -> Self {
        Self {
            metadata,
            ..self.clone()
        }
    }
}
impl<E: Clone> IPersistent for Standard<E> {}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct Mutable<E: Clone> {
    value: Standard<E>,
}
impl<E: Clone> Mutable<E> {
    pub fn new() -> Self {
        Self {
            value: Standard::new(),
        }
    }
    pub fn len(&self) -> usize {
        self.value.len()
    }
    pub fn is_empty(&self) -> bool {
        self.value.is_empty()
    }
    pub fn get(&self, index: usize) -> Option<&E> {
        self.value.get(index)
    }
    pub fn peek_first(&self) -> Option<&E> {
        self.value.peek_first()
    }
    pub fn peek_last(&self) -> Option<&E> {
        self.value.peek_last()
    }
    pub fn push_last(&mut self, value: E) -> &mut Self {
        self.value = self.value.push_last(value);
        self
    }
    pub fn pop_first(&mut self) -> &mut Self {
        self.value = self.value.pop_first_value();
        self
    }
    pub fn pop_last(&mut self) -> &mut Self {
        self.value = self.value.pop_last_value();
        self
    }
    pub fn empty(&mut self) -> &mut Self {
        self.value = Standard::new();
        self
    }
    pub fn iter(&self) -> impl Iterator<Item = &E> {
        self.value.iter()
    }
}
impl<E: Clone> FromIterator<E> for Mutable<E> {
    fn from_iter<T: IntoIterator<Item = E>>(iter: T) -> Self {
        Self {
            value: iter.into_iter().collect(),
        }
    }
}
impl<E: Clone> IMutable for Mutable<E> {}
impl<E: Clone> IToPersistent for Mutable<E> {
    type Persistent = Standard<E>;
    fn to_persistent(&mut self) -> Self::Persistent {
        self.value.clone()
    }
}
impl<E: Clone> IToMutable for Standard<E> {
    type Mutable = Mutable<E>;
    fn to_mutable(&self) -> Self::Mutable {
        Mutable {
            value: self.clone(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{Standard, MAX_LENGTH};
    #[test]
    fn crosses_head_buffer_tail_boundaries() {
        let q = (0..(MAX_LENGTH * 2 + 9)).collect::<Standard<_>>();
        assert_eq!(q.get(0), Some(&0));
        assert_eq!(q.get(MAX_LENGTH), Some(&MAX_LENGTH));
        assert_eq!(q.peek_last(), Some(&(MAX_LENGTH * 2 + 8)));
        let p = q.pop_first_value();
        assert_eq!(p.peek_first(), Some(&1));
        assert_eq!(q.peek_first(), Some(&0));
    }
    #[test]
    fn pop_last_preserves_previous() {
        let q = (0..1030).collect::<Standard<_>>();
        let p = q.pop_last_value();
        assert_eq!(q.peek_last(), Some(&1029));
        assert_eq!(p.peek_last(), Some(&1028));
    }
    #[test]
    fn persistent_operations_preserve_metadata() {
        use crate::lang::protocol::{IEmpty, IMetadata};
        let queue = Standard::from_iter(0..(MAX_LENGTH + 2))
            .with_meta(Some(crate::lang::data::Metadata::document("doc")));
        let values = [
            queue.push_last(9999),
            queue.pop_first_value(),
            queue.pop_last_value(),
            queue.empty(),
        ];
        assert!(values
            .iter()
            .all(|value| value.meta().map(|m| m.doc().unwrap()) == Some("doc")));
    }

    #[test]
    fn mutable_round_trip_crosses_chunk_boundaries() {
        use crate::lang::protocol::{IToMutable, IToPersistent};
        let original = (0..(MAX_LENGTH + 3)).collect::<Standard<_>>();
        let mut mutable = original.to_mutable();
        mutable.pop_first().pop_last().push_last(9999);
        assert_eq!(mutable.peek_first(), Some(&1));
        assert_eq!(mutable.peek_last(), Some(&9999));
        let persistent = mutable.to_persistent();
        assert_eq!(persistent.peek_first(), Some(&1));
        assert_eq!(persistent.peek_last(), Some(&9999));
        assert_eq!(original.peek_first(), Some(&0));
    }
}
